go-controller/pkg/node/base_node_network_controller.go

```diff
@@ -0,0 +1,227 @@
+
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+)
+
+// nodeUpgrade upgrades the node to the latest topology version
+func (bnnc *BaseNodeNetworkController) nodeUpgrade(ctx context.Context, isOvnUpEnabled bool,
+	cniServer *cni.Server, mgmtPortConfig *managementPortConfig, bridgeName string) error {
+	// in cases other than the default network full mode, check if isOvnUpEnabled needs to be
+	// upgraded, if not, return directly
+	if (isOvnUpEnabled || config.OvnKubeNode.DisableOVNIfaceIdVer) &&
+		!(config.OvnKubeNode.Mode == types.NodeModeFull && !bnnc.IsSecondary()) {
+		return nil
+	}
+
+	// Upgrade for Node. If we upgrade workers before masters, then we need to keep service routing via
+	// mgmt port until masters have been updated and modified OVN config. Run a goroutine to handle this case
+	upgradeController := upgrade.NewController(bnnc.client, bnnc.watchFactory)
+	initialTopoVersion, err := upgradeController.GetTopologyVersion(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get initial topology version: %w", err)
+	}
+	klog.Infof("Current control-plane topology version is %d", initialTopoVersion)
+
+	if config.OvnKubeNode.Mode == types.NodeModeFull && !bnnc.IsSecondary() {
+		needLegacySvcRoute := true
+		if (initialTopoVersion >= types.OvnHostToSvcOFTopoVersion && config.GatewayModeShared == config.Gateway.Mode) ||
+			(initialTopoVersion >= types.OvnRoutingViaHostTopoVersion) {
+			// Configure route for svc towards shared gw bridge
+			// Have to have the route to bridge for multi-NIC mode, where the default gateway may go to a non-OVS interface
+			if err := configureSvcRouteViaBridge(bridgeName); err != nil {
+				return err
+			}
+			needLegacySvcRoute = false
+		}
+
+		// Determine if we need to run upgrade checks
+		if initialTopoVersion != types.OvnCurrentTopologyVersion {
+			if needLegacySvcRoute {
+				klog.Info("System may be upgrading, falling back to legacy K8S Service via management port")
+				// add back legacy route for service via management port
+				link, err := util.LinkSetUp(types.K8sMgmtIntfName)
+				if err != nil {
+					return fmt.Errorf("unable to get link for %s, error: %v", types.K8sMgmtIntfName, err)
+				}
+				var gwIP net.IP
+				for _, subnet := range config.Kubernetes.ServiceCIDRs {
+					if utilnet.IsIPv4CIDR(subnet) {
+						gwIP = mgmtPortConfig.ipv4.gwIP
+					} else {
+						gwIP = mgmtPortConfig.ipv6.gwIP
+					}
+					err := util.LinkRoutesApply(link, gwIP, []*net.IPNet{subnet}, config.Default.RoutableMTU, nil)
+					if err != nil {
+						return fmt.Errorf("unable to add legacy route for services via mp0, error: %v", err)
+					}
+				}
+			}
+		}
+	}
+
+	// need to run upgrade controller
+	go func() {
+		if err := upgradeController.WaitForTopologyVersion(ctx, bnnc.stopChan, types.OvnCurrentTopologyVersion, 30*time.Minute); err != nil {
+			if !errors.Is(err, upgrade.ErrStopped) {
+				klog.Fatalf("Error while waiting for Topology Version to be updated: %v", err)
```

Comment:
I think nodeUpgrade should return the error and let the caller decide what to do with it.

Reply:
for default network, it is fatal. but for secondary network, we could just log an error without doing anything. 

This is a go routine, so at time of failure, watchPodsDPU() may be already running, and I don't see a need of retrying.

Reply:
If it fails for a secondary network and we don't retry, it would mean that that secondary network would have not process its topology upgrade. Right now this would mean for example not using isOVNUpEnabled. So there is a consequence, why would we not retry? If we retried and were able to update isOVNUpEnabled, watchPodsDPU would use the new value?

BTW this is another reason why doing this globally looks better. At least the `waitForTopologyVersion` part.

Reply:
how to retry? Start() already returned, and what is left are all go routines (worker thread). Or we need to introduce another stopChan for those threads to watch for.

Reply:
do waitForTopologyVersion forever, with no timeout? When it eventually succeeds, isOVNUpEnabled will be updated in the secondary network controller, picked up and used by WatchPodsDPU? Isn't that the reason we are doing all this for?

Reply:
If waitForTopologyVersion() could fail in 30 minutes that means something is seriously wrong, right? default network full mode must have a timeout, I don't see why other mode need to do it differently.

if we differ the two case and call two different version of  waitForTopologyVersion() in default network full mode and other cases, this shared function would be uglier :-(